cms/djangoapps/contentstore/views/tests/test_item.py

```diff
@@ -680,6 +682,202 @@ def verify_name(source_usage_key, parent_usage_key, expected_name, display_name=
+        self.assertEqual(response.status_code, 200)
+        response = json.loads(response.content)
+        self.assertEqual(response['move_source_locator'], unicode(source_usage_key))
+        self.assertEqual(response['parent_locator'], unicode(target_usage_key))
+        self.assertEqual(response['source_index'], actual_index)
+        new_parent_loc = self.store.get_parent_location(source_usage_key)
+        self.assertEqual(new_parent_loc, target_usage_key)
+        self.assertNotEqual(parent_loc, new_parent_loc)
+
+    def test_move_component(self):
+        """
+        Test move component with different xblock types.
+        """
+        for source_usage_key, target_usage_key in [
+                (self.html_usage_key, self.vert2_usage_key),
+                (self.vert_usage_key, self.seq2_usage_key),
+                (self.seq_usage_key, self.chapter2_usage_key)
+        ]:
+            self.assert_move_item(source_usage_key, target_usage_key)
+
+    def test_move_source_index(self):
+        """
+        Test moving an item to a particular index.
+        """
+        parent = self.get_item_from_modulestore(self.vert_usage_key)
+        children = parent.get_children()
+        self.assertEqual(len(children), 2)
+
+        # Create a component within vert2.
+        resp = self.create_xblock(parent_usage_key=self.vert2_usage_key, display_name='html2', category='html')
+        html2_usage_key = self.response_usage_key(resp)
+
+        # Move html2_usage_key inside vert_usage_key at second position.
+        self.assert_move_item(html2_usage_key, self.vert_usage_key, 1)
+        parent = self.get_item_from_modulestore(self.vert_usage_key)
+        children = parent.get_children()
+        self.assertEqual(len(children), 3)
+        self.assertEqual(children[1].location, html2_usage_key)
+
+    def test_move_undo(self):
+        """
+        Test move a component and move it back (undo).
+        """
+        self.assert_move_item(self.html_usage_key, self.vert2_usage_key)
+        self.assert_move_item(self.html_usage_key, self.vert_usage_key)
+
+    def test_move_moved_parent(self):
+        """
+        Test move a moved parents.
+        """
+        self.assert_move_item(self.vert_usage_key, self.seq2_usage_key)
+        self.assert_move_item(self.seq2_usage_key, self.chapter2_usage_key)
+
+        # check vert_usage_key has seq2_usage_key as parent
+        self.assertEqual(self.store.get_parent_location(self.vert_usage_key), self.seq2_usage_key)
+
+    def test_move_large_source_index(self):
+        """
+        Test moving an item at a large index would insert item into target location at the end..
+        """
+        parent = self.get_item_from_modulestore(self.vert2_usage_key)
+        parent_children_length = len(parent.children)
+        response = self._move_component(self.html_usage_key, self.vert2_usage_key, parent_children_length + 10)
```

Comment:
IMO it would be better to return an error for a target index that doesn't exist. It seems that it could introduce subtle bugs where the math was wrong and then items ended up incorrectly at the end.